zboxcore/sdk/dirworker.go

```diff
@@ -4,66 +4,123 @@ import (
 	"bytes"
 	"context"
 	"errors"
+	"fmt"
 	"io/ioutil"
+	"math/bits"
 	"mime/multipart"
 	"net/http"
-	"path/filepath"
 	"strings"
+	"sync"
 	"time"
 
+	"github.com/0chain/gosdk/zboxcore/allocationchange"
 	"github.com/0chain/gosdk/zboxcore/blockchain"
+	"github.com/0chain/gosdk/zboxcore/client"
 	l "github.com/0chain/gosdk/zboxcore/logger"
 	"github.com/0chain/gosdk/zboxcore/zboxutil"
 )
 
+const (
+	DirectoryExists = "directory_exists"
+)
+
 type DirRequest struct {
 	allocationID string
-	name         string
+	allocationTx string
+	remotePath   string
+	blobbers     []*blockchain.StorageNode
 	ctx          context.Context
-	action       string // create, del
+	wg           *sync.WaitGroup
+	dirMask      uint32
+	mu           *sync.Mutex
 	connectionID string
 	Consensus
 }
 
 func (req *DirRequest) ProcessDir(a *Allocation) error {
 	numList := len(a.Blobbers)
+	req.wg = &sync.WaitGroup{}
```

Comment:
@peterlimg I can initi wg and use it but don't understand how it will cause race condition?

Reply:
If any chance the DirRequest.ProcessDir will be called concurrently, the req.wg will have race condition issue. If it would never be used concurrently, we'd better not define it as a member variable as it may confuse users. So let's just use it as a temporary var.